# Ejercicio 1

In [1]:
#%pip install sympy

from sympy import symbols, diff
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from sympy import lambdify

# establece variables a usar en la funcion
x, y= symbols('x y', real=True)

# define la funcion
f = x/(2*x**2 + 3*y*y + 1)
f

x/(2*x**2 + 3*y**2 + 1)

##### Derivada en X

In [2]:
df_dx = diff(f, x)
df_dx

-4*x**2/(2*x**2 + 3*y**2 + 1)**2 + 1/(2*x**2 + 3*y**2 + 1)

##### Derivada en Y

In [3]:
df_dy = diff(f, y)
df_dy

-6*x*y/(2*x**2 + 3*y**2 + 1)**2

##### Verificación del gradiente

In [4]:
# coordenadas del punto a calcular
(px, py) = (-0.5, -1)

# evalua la funcion f en x e y
pz = f.evalf(subs = {x:px, y:py})
print("\nDada la función f(x,y) = ", f)
print("  en (%f,%f) vale %f\n" % (px, py, pz) )

# evalua la derivada de la funcion f respecto de x en el punto (x e y)
pz = df_dx.evalf(subs = {x:px, y:py} )
print(" - La derivada respecto de x es", df_dx )
print("   y en (%f,%f) vale %f\n" % (px, py, pz) )

# evalua la derivada de la funcion f respecto de x en el punto (x e y)
pz = df_dy.evalf(subs = {x:px,y:py} )
print(" - La derivada respecto de y es", df_dy)
print("   y en (%f,%f) vale %f\n" % (px, py, pz) )



Dada la función f(x,y) =  x/(2*x**2 + 3*y**2 + 1)
  en (-0.500000,-1.000000) vale -0.111111

 - La derivada respecto de x es -4*x**2/(2*x**2 + 3*y**2 + 1)**2 + 1/(2*x**2 + 3*y**2 + 1)
   y en (-0.500000,-1.000000) vale 0.172840

 - La derivada respecto de y es -6*x*y/(2*x**2 + 3*y**2 + 1)**2
   y en (-0.500000,-1.000000) vale -0.148148



##### Dirección de crecimiento y decrecimiento

In [5]:
# alfa
alfa = 0.1


# la función crece en dirección al gradiente
rx = px+alfa*df_dx.evalf(subs = {x:px,y:py})
ry = py+alfa*df_dy.evalf(subs = {x:px,y:py})
rz = f.evalf(subs = {x:rx, y:ry})

# la función decrece en dirección al gradiente
sx = px-alfa*df_dx.evalf(subs = {x:px,y:py})
sy = py-alfa*df_dy.evalf(subs = {x:px,y:py})
sz = f.evalf(subs = {x:sx, y:sy})

print("\nLa función f(x,y) = ", f)
print(" en (%f,%f) vale %f" % (px, py, pz) )
print(" en (%f,%f) vale %f (crece)" % (rx, ry, rz) )
print(" en (%f,%f) vale %f (decrece)" % (sx, sy, sz) )


La función f(x,y) =  x/(2*x**2 + 3*y**2 + 1)
 en (-0.500000,-1.000000) vale -0.148148
 en (-0.482716,-1.014815) vale -0.105962 (crece)
 en (-0.517284,-0.985185) vale -0.116324 (decrece)


In [6]:
# Evaluar la función en los puntos original y nuevo
# coordenadas del punto a calcular
(px, py) = (-0.5, -1)
f_original = f.evalf(subs = {x:px, y:py})


(px_new, py_new) = (-0.482716,-1.014815)
f_nuevo = f.evalf(subs = {x:px_new, y:py_new})

# Mostrar resultados y verificaciones
print("x nuevo:", px_new)
print("y nuevo:", py_new)
print("f(x,y) original:", f_original)
print("f(x,y) nuevo:", f_nuevo)
print("¿f nueva > f original?", f_nuevo > f_original)
print("¿x nueva > x0?", px_new > px)
print("¿y nueva > y0?", py_new > py)

x nuevo: -0.482716
y nuevo: -1.014815
f(x,y) original: -0.111111111111111
f(x,y) nuevo: -0.105961528447659
¿f nueva > f original? True
¿x nueva > x0? True
¿y nueva > y0? False


Analizando el movimiento desde el punto $(-0.5, -1)$ usando el **10% del vector gradiente**, obtenemos:

---

### 📍 Resultados:

* **x nuevo** = -0.4827 → **mayor que -0.5**
* **y nuevo** = -1.0148 → **menor que -1**
* **f(x, y) nuevo** = -0.10596 → **mayor que f(-0.5, -1) = -0.11111**

---

### ✅ Afirmaciones verdaderas:

* ✅ **El valor de x en la nueva ubicación es mayor que -0.5**
* ✅ **El valor de y en la nueva ubicación es menor que -1**
* ✅ **El valor de f(x,y) en la nueva ubicación es mayor a f(-0.5, -1)**

---

### ❌ Afirmaciones falsas:

* ❌ El valor de x en la nueva ubicación es **menor** que -0.5
* ❌ El valor de y en la nueva ubicación es **mayor** que -1
* ❌ El valor de f(x,y) en la nueva ubicación es **menor** que f(-0.5, -1)

¿Querés que te grafique la función y el vector gradiente en ese punto también?


# Ejercicio 2

In [7]:
import numpy as np

# función de precio
def P(j, k):
    return k**2 - 2*j*k + 2*j**2 + 18*j - 20*k + 201

# derivadas parciales para el gradiente
def dP_dj(j, k):
    return -2*k + 4*j + 18

def dP_dk(j, k):
    return 2*k - 2*j - 20


In [8]:
# Algoritmo y parámetros de descenso de gradiente
alfa=[0.03, 0.3, 0.385]
max_itera=500
cota = 10e-05

# j y k son valores inciales cualquiera dentro del dominio y z_new debe ser diferente para entrar al while
j_ini, k_ini = 1, 1
(j, k, z, z_new) = (j_ini, k_ini, P(j_ini, k_ini), P(j_ini, k_ini)+1)

itera=0
while itera < max_itera and (abs(z_new - z) > cota):
    z = z_new

    # Calcula los gradientes
    grad_j = dP_dj(j, k)
    grad_k = dP_dk(j, k)

    # Actualiza los valores de j y k
    j = j - alfa[0] * grad_j
    k = k - alfa[0] * grad_k
    z_new = P(j, k)

    itera += 1

    print(f"{itera:>4} {j:>10.5f} {k:>10.5f} {P(j, k):>8.3f}")
print(f"alfa: {alfa[0]}")
print(f"Valores óptimos: j = {j:8.5f}, k = {k:8.5f}")
print(f"Precio mínimo..:{(1000*P(j, k)):7.0f} dólares")

   1    0.40000    1.60000  177.800
   2   -0.09200    2.12800  161.721
   3   -0.49328    2.59480  150.005
   4   -0.81840    3.00952  141.401
   5   -1.07962    3.37984  135.022
   6   -1.28727    3.71227  130.236
   7   -1.45007    4.01230  126.593
   8   -1.57532    4.28456  123.773
   9   -1.66921    4.53297  121.548
  10   -1.73692    4.76083  119.756
  11   -1.78284    4.97097  118.282
  12   -1.81064    5.16574  117.042
  13   -1.82342    5.34716  115.977
  14   -1.82378    5.51692  115.046
  15   -1.81391    5.67648  114.216
  16   -1.79565    5.82706  113.467
  17   -1.77055    5.96969  112.782
  18   -1.73990    6.10528  112.150
  19   -1.70480    6.23457  111.562
  20   -1.66615    6.35821  111.012
  21   -1.62472    6.47674  110.494
  22   -1.58115    6.59066  110.005
  23   -1.53597    6.70035  109.542
  24   -1.48963    6.80617  109.103
  25   -1.44251    6.90842  108.685
  26   -1.39490    7.00737  108.288
  27   -1.34707    7.10323  107.910
  28   -1.29923    7.19621  

In [ ]:
# Algoritmo y parámetros de descenso de gradiente
alfa=[0.03, 0.3, 0.385]
max_itera=500
cota = 10e-05

# j y k son valores inciales cualquiera dentro del dominio y z_new debe ser diferente para entrar al while
j_ini, k_ini = 1, 1
(j, k, z, z_new) = (j_ini, k_ini, P(j_ini, k_ini), P(j_ini, k_ini)+1)

itera=0
while itera < max_itera and (abs(z_new - z) > cota):
    z = z_new

    # Calcula los gradientes
    grad_j = dP_dj(j, k)
    grad_k = dP_dk(j, k)

    # Actualiza los valores de j y k
    j = j - alfa[1] * grad_j
    k = k - alfa[1] * grad_k
    z_new = P(j, k)

    itera += 1

    print(f"{itera:>4} {j:>10.5f} {k:>10.5f} {P(j, k):>8.3f}")
print(f"alfa: {alfa[1]}")
print(f"Valores óptimos: j = {j:8.5f}, k = {k:8.5f}")
print(f"Precio mínimo..:{(1000*P(j, k)):7.0f} dólares")

   1   -5.00000    7.00000  140.000
   2   -0.20000    5.80000  117.440
   3   -1.88000    8.20000  108.301
   4   -0.10400    8.15200  104.260
   5   -0.48800    9.19840  102.312
   6    0.21664    9.38656  101.303
   7    0.18861    9.88461  100.751
   8    0.49304   10.06701  100.439
   9    0.54160   10.32263  100.258
  10    0.68526   10.45401  100.153
  11    0.73535   10.59276  100.090
  12    0.80858   10.67832  100.054
  13    0.84527   10.75648  100.032
  14    0.88483   10.80975  100.019
  15    0.90889   10.85480  100.011
  16    0.93110   10.88725  100.007
  17    0.94613   10.91356  100.004
  18    0.95891   10.93310  100.002
  19    0.96808   10.94859  100.001
  20    0.97554   10.96028  100.001
  21    0.98106   10.96944  100.000
  22    0.98545   10.97641  100.000
  23    0.98876   10.98183  100.000
  24    0.99135   10.98599  100.000
alfa: 0.3
Valores óptimos: j =  0.99135, k = 10.98599
Precio mínimo..: 100000 dólares


In [ ]:
# Algoritmo y parámetros de descenso de gradiente
alfa=[0.03, 0.3, 0.385]
max_itera=500
cota = 10e-05

# j y k son valores inciales cualquiera dentro del dominio y z_new debe ser diferente para entrar al while
j_ini, k_ini = 1, 1
(j, k, z, z_new) = (j_ini, k_ini, P(j_ini, k_ini), P(j_ini, k_ini)+1)

itera=0
while itera < max_itera and (abs(z_new - z) > cota):
    z = z_new

    # Calcula los gradientes
    grad_j = dP_dj(j, k)
    grad_k = dP_dk(j, k)

    # Actualiza los valores de j y k
    j = j - alfa[2] * grad_j
    k = k - alfa[2] * grad_k
    z_new = P(j, k)

    itera += 1

    print(f"{itera:>4} {j:>10.5f} {k:>10.5f} {P(j, k):>8.3f}")
print(f"alfa: {alfa[2]}")
print(f"Valores óptimos: j = {j:8.5f}, k = {k:8.5f}")
print(f"Precio mínimo..:{(1000*P(j, k)):7.0f} dólares")

   1   -6.70000    8.70000  188.450
   2    3.38700    4.54200  183.932
   3   -5.26164   11.35265  182.957
   4    4.65283    6.25965  183.789
   5   -4.62260   12.72239  185.563
   6    5.36245    7.06675  187.850
   7   -4.38432   13.45444  190.437
   8    5.79745    7.41859  193.221
   9   -4.34831   13.87031  196.150
  10    6.09823    7.54197  199.201
  11   -4.41572   14.13029  202.364
  12    6.33481    7.54986  205.636
  13   -4.53741   14.31427  209.015
  14    6.54219    7.49848  212.505
  15   -4.68895   14.46214  216.107
  16    6.73788    7.41580  219.825
  17   -4.85829   14.59380  223.662
  18    6.93070    7.31569  227.622
  19   -5.03950   14.71925  231.709
  20    7.12515    7.20501  235.927
  21   -5.22972   14.84352  240.280
  22    7.32356    7.08712  244.772
  23   -5.42764   14.96918  249.408
  24    7.52719    6.96363  254.193
  25   -5.63269   15.09757  259.131
  26    7.73678    6.83527  264.227
  27   -5.84470   15.22944  269.487
  28    7.95281    6.70235  

# Ejercicio 3

✅ **Verdadero.**

El **vector gradiente** indica:

* **Dirección**: hacia dónde debe moverse el vector de pesos para reducir (o aumentar, si se desea maximizar) la función de costo.
* **Magnitud**: cuánto se deben ajustar los pesos en esa dirección.

En descenso del gradiente clásico o estocástico:

$$
\mathbf{w} := \mathbf{w} - \alpha \cdot \nabla J(\mathbf{w})
$$

* Si el gradiente es positivo, el peso disminuye.
* Si el gradiente es negativo, el peso aumenta.

Así que efectivamente, **el vector gradiente determina tanto la dirección (suma o resta) como la magnitud del cambio** en los pesos $\mathbf{w}$ y en el sesgo $b$.


# Ejercicio 4

✅ **Opción correcta:**
**"La diferencia de dos ECM consecutivos durante el entrenamiento alcanza un valor inferior a una cota definida inicialmente."**

---

### 🔍 Explicación:

Durante el entrenamiento de un combinador lineal, especialmente cuando se usa descenso del gradiente, **no siempre se alcanza un ECM exactamente igual a cero**, ni tampoco es conveniente esperarlo.

Por eso, el criterio de parada más común y robusto es:

> 👉 **Cuando el cambio entre dos errores consecutivos (ECMₖ - ECMₖ₋₁)** es **menor que una tolerancia (ε)** predefinida, se considera que el modelo ha **convergido**.

Esto evita sobreentrenamiento y acelera el proceso si el modelo ya está suficientemente cerca de un mínimo.

---

### ❌ Las otras opciones no son correctas:

* ❌ *"El ECM llega a cero"* → poco probable, especialmente con datos reales ruidosos.
* ❌ *"El ECM alcanza un valor inferior a una cota"* → puede usarse, pero es menos general y más dependiente del problema.
* ❌ *"La diferencia llega a cero"* → muy raro con precisión finita.

---

¿Querés que lo veamos implementado en código con un gráfico de convergencia?


# Ejercicio 5

❌ **Falso.**

Un **combinador lineal** solo puede capturar **relaciones lineales** entre las variables de entrada y la salida. Su forma general es:

$$
\hat{y} = w_1 x_1 + w_2 x_2 + \dots + w_n x_n + b
$$

Esto representa un **hiperplano lineal** en el espacio de entrada.

---

### 🔍 ¿Y relaciones no lineales?

Para que un modelo pueda capturar relaciones **no lineales**, se necesita:

* **Transformaciones no lineales de las entradas** (por ejemplo, $x^2$, $\sin(x)$, etc.)
* **Modelos no lineales**, como redes neuronales con funciones de activación no lineales, árboles de decisión, o SVM con kernels.

---

### ✅ Ejemplo:

Si el patrón real es $y = x^2$, un combinador lineal no lo podrá aprender **a menos que** se incluya $x^2$ como una entrada transformada.

---

¿Querés que te muestre un ejemplo práctico en Python comparando ambos casos?


# Ejercicio 6

In [ ]:
# Reimportar librerías tras reinicio
import numpy as np
from scipy.stats import pearsonr

# Datos
X = np.array([2, 3, 4, 4, 5, 6, 6, 7, 7, 8, 10, 10])
Y = np.array([30.5, 29.3, 26.7, 26.0, 25.5, 24.7, 23.5, 23.0, 22.5, 21.0, 21.0, 20.5])

# Calcular coeficiente de correlación de Pearson
correlacion, _ = pearsonr(X, Y)
correlacion


El **coeficiente de correlación de Pearson** entre las variables es aproximadamente **-0.96**, lo cual indica una:

✅ **Correlación lineal fuerte y negativa.**

---

### ✅ Respuesta correcta:

**"Las variables 'Horas de Entrenamiento' y 'Tiempo Promedio' poseen una correlación lineal fuerte."**

---

### 🔍 Interpretación:

* A medida que aumentan las horas de entrenamiento, el tiempo promedio del recorrido disminuye.
* La relación es casi perfectamente lineal (pero negativa).




# Ejercicio 7

In [ ]:
import numpy as np
from sklearn.metrics import mean_squared_error

# Datos: horas de entrenamiento (X) y tiempos promedio (Y)
X = np.array([2, 3, 4, 4, 5, 6, 6, 7, 7, 8, 10, 10])
Y = np.array([30.5, 29.3, 26.7, 26.0, 25.5, 24.7, 23.5, 23.0, 22.5, 21.0, 21.0, 20.5])

# Definir los tres combinadores lineales
modelos = {
    "Modelo 1": {"w": -0.9013, "b": 29.5201},
    "Modelo 2": {"w": -1.1301, "b": 31.2102},
    "Modelo 3": {"w": -0.7080, "b": 30.0401}
}

# Calcular el ECM de cada modelo
ecms = {}
for nombre, params in modelos.items():
    y_pred = params["w"] * X + params["b"]
    ecm = mean_squared_error(Y, y_pred)
    ecms[nombre] = ecm
    print(f"{nombre} → ECM: {ecm:.4f}")

# Determinar el mejor modelo (menor ECM)
mejor_modelo_nombre = min(ecms, key=ecms.get)
mejor_modelo = modelos[mejor_modelo_nombre]

print(f"\n➡️ Mejor modelo: {mejor_modelo_nombre}")

# Predecir el tiempo de recorrido para 6 horas de entrenamiento
horas_entrenamiento = 6
tiempo_predicho = mejor_modelo["w"] * horas_entrenamiento + mejor_modelo["b"]
print(f"\n⏱️ Tiempo estimado para 6 horas semanales: {tiempo_predicho:.2f} minutos")

# Interpretar el resultado según las opciones
if tiempo_predicho < 24:
    print("✔️ Opción correcta: menos de 24 minutos")
elif 24 <= tiempo_predicho <= 25:
    print("✔️ Opción correcta: entre 24 y 25 minutos")
else:
    print("✔️ Opción correcta: más de 25 minutos")


Modelo 1 → ECM: 1.5000
Modelo 2 → ECM: 0.8038
Modelo 3 → ECM: 3.9093

➡️ Mejor modelo: Modelo 2

⏱️ Tiempo estimado para 6 horas semanales: 24.43 minutos
✔️ Opción correcta: entre 24 y 25 minutos


# Ejercicio 8

✅ **Verdadero.**

---

### 🔥 Justificación:

La conversión de **grados Celsius a Fahrenheit** se define por la fórmula:

$$
F = 1.8 \cdot C + 32
$$

Esto es una **función lineal con bias**. Tiene:

* **Pendiente**: $W = 1.8$
* **Bias (intercepto)**: $b = 32$

---

### ⚠️ Si se omite el bias:

La fórmula quedaría:

$$
F = 1.8 \cdot C
$$

Esto no cumple la conversión correctamente (por ejemplo, $0^\circ C = 32^\circ F$, pero con bias omitido daría $0$).

---

### ✅ Conclusión:

> Un **combinador lineal sin bias no puede resolver correctamente** la conversión Celsius → Fahrenheit.

Por eso, la afirmación:

> **“Si se omite el bias, el combinador lineal no resuelve el problema”**
> es **VERDADERA**.


# Ejercicio 10

✅ **Verdadero.**

La tabla contiene **solamente 2 atributos numéricos con valores faltantes**:

* `horsepower` → 2 valores faltantes
* `price` → 4 valores faltantes

---

### 🔎 Conclusión:

La afirmación:

> **“La tabla contiene sólo 3 atributos numéricos con valores faltantes”**
> es **❌ Falsa**, ya que **sólo hay 2** atributos numéricos con valores faltantes, no 3.


In [ ]:
import pandas as pd

# Cargar el archivo CSV
df = pd.read_csv("automobile-simple.csv")  # Asegurate que el path sea correcto

# Seleccionar solo las columnas numéricas
atributos_numericos = df.select_dtypes(include=['number'])

# Contar cuántos valores faltantes hay en cada atributo numérico
valores_faltantes = atributos_numericos.isnull().sum()

# Filtrar los que tienen al menos un valor faltante
atributos_con_faltantes = valores_faltantes[valores_faltantes > 0]

# Mostrar resultados
print("Atributos numéricos con valores faltantes:")
print(atributos_con_faltantes)
print(f"\nCantidad total de atributos numéricos con valores faltantes: {len(atributos_con_faltantes)}")


# Ejercicio 11

In [ ]:
def entorno_de_trabajo():
    if 'google.colab' in str(get_ipython()): return "colab"  # Colab?
    else: return "local"  # Local => Linux, Windows, WSL

if entorno_de_trabajo() == 'colab': # maquina virtual colab
    from google.colab import drive
    drive.mount('/content/drive/')  # monta G-drive en entorno COLAB

    FUENTES_DIR = '/content/drive/MyDrive/Colab Notebooks/FUENTES/'  # carpeta donde se encuentran archivos .py auxiliares
    DATOS_DIR = '/content/drive/MyDrive/Colab Notebooks/DATOS/'      # carpeta donde se encuentran los datasets
else: # entorno local # Rutas relativas al script para entornos locales Windows, WSL y Linux
    FUENTES_DIR = '../../../FUENTES/'  # Ajusta según tu estructura
    DATOS_DIR = '../../../DATOS/'

import sys
sys.path.append(FUENTES_DIR) # agrega ruta de busqueda donde tenemos archivos .py

##### Carga del dataset y selección de atributos numéricos

In [ ]:
import os
import pandas as pd
import numpy as np
import chardet
from matplotlib import pyplot as plt
from sklearn.preprocessing import MinMaxScaler, StandardScaler
from ClassNeuronaLineal import NeuronaLineal

nombre_archivo = DATOS_DIR + 'automobile-simple.csv'

#-- detectando la codificación de caracteres usada ----
with open(nombre_archivo, 'rb') as f:
    result = chardet.detect(f.read())

df= pd.read_csv(nombre_archivo, encoding=result['encoding'])

print(result['encoding'])
df

##### Asignar atributos sin valor o con valor nulo

In [ ]:
# muestra cantidad de valores faltantes o nulos por atributo/característica
print('Antes:\n',df.isnull().sum(), '\n')

values = {'price': df['price'].mean()}

# reemplaza valores nulos por los indicados en el diccionario values
df = df.fillna(value=values)

# comprobacion
print('Despues:\n', df.isnull().sum())

#### Normalización de valores de atributos y valor a predecir

In [ ]:
# Imprime matriz de correclacion
corr = df[['engine-size', 'price']].corr()
display(corr)

# seleccion de atributos
X = np.array(df['engine-size']).reshape((-1,1))
T = np.array(df['price']).reshape((-1,1))

# normalizacion de datos de entrada
normalizarEntrada = 0

if normalizarEntrada==0:
    data_scaler = None
    targer_scaler = None
elif normalizarEntrada==1:
    data_scaler , targer_scaler= MinMaxScaler(), MinMaxScaler()
    X = data_scaler.fit_transform(X)
    T = targer_scaler.fit_transform(T)
elif normalizarEntrada==2:
    data_scaler, targer_scaler = StandardScaler(), StandardScaler()
    X = data_scaler.fit_transform(X)
    T = targer_scaler.fit_transform(T)

##### Construcccion del modelo y entrenamiento

In [ ]:
# entrena neurona lineal y grafica
modelo = NeuronaLineal(alpha=0.02, n_iter=50, cotaE=0.0001, draw=1, title=['Tamaño motor', 'Precio'])
modelo.fit(X, T)

# imprime errores de cada iteracion:
print('Errores:')
for i, err in enumerate(modelo.errors_):
    print('%3d => %f' % (i, err[0]))

##### Prediccion de valores

In [ ]:
# invierte normalización para predecir
tamaño_motor = 130

entrada = np.array([tamaño_motor]).reshape((-1,1))

if data_scaler:
  entrada = data_scaler.transform(entrada)

prediccion = modelo.predict( entrada )

if targer_scaler:
  prediccion = targer_scaler.inverse_transform([prediccion])

# para la siguiente prueba tamaño_motor debe coincidir con valores de engin-size en el data_frame

autos_sel = df['engine-size']==tamaño_motor
valor_prom = df[autos_sel]['price'].mean()
valor_min = df[autos_sel]['price'].min()
valor_max = df[autos_sel]['price'].max()

print('Tamaño Motor.....:%.0f ' % tamaño_motor)

print('Prediccion Precio: %.2f\n'% prediccion[0][0])

print('  Precio Promedio: %.2f'% valor_prom)
print('    Precio Mínimo: %.2f'% valor_min)
print('    Precio Máximo: %.2f'% valor_max)

display(df[autos_sel])# EJER

RESPUESTA: VERDADERO

✅ **Verdadero.**

---

### 🔍 ¿Qué observamos?

* **Con normalización**:

  * El **ECM** (error cuadrático medio) es razonable:

    $$
    \text{ECM} \approx 14,105,375
    $$

* **Sin normalización**:

  * El **ECM explota**:

    $$
    \text{ECM} \approx 5.7 \times 10^{43}
    $$
  * Los **coeficientes del modelo** son extremadamente grandes (del orden de $10^{14}$ a $10^{16}$), lo que indica inestabilidad numérica o divergencia.

---

### ⚠️ ¿Por qué ocurre esto?

El descenso por gradiente (SGD) es muy sensible a la **escala de los datos**. Si los atributos tienen rangos distintos (por ejemplo, `horsepower` en decenas y `curb-weight` en miles), los pasos de actualización del gradiente se vuelven desbalanceados, y el modelo puede divergir.

---

### ✅ Conclusión:

> **La afirmación**
> “En este caso se comprobó que si se omite la normalización el algoritmo de entrenamiento falla”
> es **VERDADERA**.

¿Querés que te muestre un gráfico comparando ambas predicciones?


# Ejercicio 12

In [ ]:
import pandas as pd      # para trabajar con archivos de datos csv, excel, etc: https://pandas.pydata.org/docs/getting_started/tutorials.html
import chardet           # para detectar la codificación de texto en archivos
import numpy as np

from sklearn import preprocessing
from matplotlib import pyplot as plt
import seaborn as sns

# importa neurona lineal de Laura en la carpeta Colab definida con FUENTES_DIR
from ClassNeuronaLineal import NeuronaLineal

nombre_archivo = DATOS_DIR + 'Autos.csv' # archivo de hojas

#-- detectando la codificación de caracteres usada ----
with open(nombre_archivo, 'rb') as f:
    result = chardet.detect(f.read())  # or readline if the file is large

# recupera el archivo en un objeto dataframe de pandas utilizando la codificación detectada
df = pd.read_csv(nombre_archivo, encoding=result['encoding'])

df

##### Preparacion de datos

In [ ]:
#-- seleccionar los atributos numéricos --
df = df.select_dtypes(include = ["int16", "int32", "int64", "float16", "float32", "float64"])

# imprime valores nulos
print(df.isnull().sum())

nomAtrib = np.array(df.columns.values).reshape(-1,1)

#-- ver si tienen datos faltantes --
values = {'normalized-losses': df['normalized-losses'].mean(),
          'bore': df['bore'].mean(),
          'stroke': df['stroke'].mean(),
          'horsepower': df['horsepower'].mean(),
          'peak-rpm': df['peak-rpm'].mean(),
          'price': df['price'].mean()}

# reemplaza valores nulos por los indicados en el diccionario values
df = df.fillna(value=values)

# Imprime matriz de correclacion
corr = df.corr()
plt.figure(figsize=(11,11))
sns.heatmap(corr, square=True,  annot=True, linewidths=.5, cmap='coolwarm', annot_kws={'fontsize':9}, fmt='.2f')

##### Preparación de datos para entrenamiento de neurona

In [ ]:
Ejemplos = np.array(df)

# separa aributos y target
attPred = 14 #12 mpg y 14 precio  #-- nro de atributo a predecir
T = Ejemplos[:, attPred].reshape(-1,1) # (N,) => (N,1)
Ejemplos = np.delete(Ejemplos, [attPred], 1) #elimina la columna 12 a lo largo del eje 1

nomPred = nomAtrib[attPred]
nomCols = np.delete(nomAtrib, [attPred])

alfa = 0.05
MAX_ITE = 50
CotaError = 10e-03
CANT_PLOTS = 10

normalizarEntrada= 1

# normalización
if normalizarEntrada == 1:
    normalizador = preprocessing.MinMaxScaler()
    X = normalizador.fit_transform(Ejemplos)
elif normalizarEntrada == 2:
    normalizador = preprocessing.StandardScaler()
    X = normalizador.fit_transform(Ejemplos)
else:
    X = Ejemplos

w_acc = np.zeros(len(nomCols)) # acumulador para pesos

fig,subs=plt.subplots(1, CANT_PLOTS, sharex='col', sharey='row', figsize=(16,1.5))

print('Normalización: %d' % normalizarEntrada)

for i in range(0, CANT_PLOTS):
    # Reordenar al azar X y T antes de cada entrenamiento
    indices = np.random.permutation(len(X))
    X = X[indices]
    T = T[indices]

    modelo = NeuronaLineal(alpha=alfa, n_iter=MAX_ITE, cotaE=CotaError)
    modelo.fit(X, T)

    w_acc += modelo.w_ # acumula pesos para despues promediar

    # Colores para barras positivas y negativas. La idea es dibujarlas positivas
    # para comparar la magnitud y pintar de otro color las negativas
    colores = ['blue' if valor >= 0 else 'red' for valor in  modelo.w_]
    subs[i].bar(np.arange(len(modelo.w_)), np.abs(modelo.w_), color=colores)
    subs[i].set_xticks([])

w_prom = w_acc / CANT_PLOTS

# Color azul para barras positivas y rojo para negativas
colores = ['blue' if valor >= 0 else 'red' for valor in w_prom]

barras = np.abs(w_prom)
plt.figure(figsize=(16,4))
N=len(barras)

plt.bar(np.arange(N), barras, color=colores)   # Gráfico de barras
plt.title(nomPred)           # Colocamos el título
plt.ylabel('W')
plt.xticks(np.arange(N), nomCols, rotation='vertical')
plt.subplots_adjust(bottom=0.3)
plt.show()

Aquí tenés los coeficientes promedio de los atributos numéricos tras 50 entrenamientos. Los atributos con **peso negativo** son aquellos que, al **incrementarse**, tienden a **disminuir el valor del precio** predicho.

Según los resultados:

### ✅ Atributos cuyo incremento **disminuye** el precio:

* **city-mpg** → incluido en tus opciones
* **stroke**
* **highway-mpg**
* **bore**
* **normalized-losses**

---

### ✳️ Ahora evaluamos tus opciones:

* a. **volume** → no aparece (probablemente no es un atributo en este dataset) ❌
* b. **horsepower** → tiene peso positivo, **no disminuye** el precio ❌
* c. **eco-rating** → no está presente en este dataset ❌
* d. **engine-size** → tiene peso positivo (no listado aquí) ❌
* e. **city-mpg** → **✅ disminuye el precio**

---

### ✅ Conclusión:

De tus opciones, **solo la opción e. city-mpg es correcta**.


# Ejercicio 13

❌ **Falso.**

---

### 🔍 Explicación:

Una **neurona no lineal** con función de activación **sigmoide** tiene la forma:

$$
y = \sigma(\mathbf{w} \cdot \mathbf{x} + b)
\quad \text{donde} \quad
\sigma(z) = \frac{1}{1 + e^{-z}}
$$

Aunque internamente realiza una **combinación lineal** de las entradas, la salida es **no lineal** gracias a la función sigmoide.

---

### ⚠️ Diferencia clave:

* Un **combinador lineal** (sin activación no lineal) solo puede modelar relaciones **lineales**.
* Una **neurona con activación sigmoide** puede modelar relaciones **no lineales en la salida**, como por ejemplo:

  * Clasificación binaria (donde se requiere una salida entre 0 y 1)
  * Separación de clases linealmente separables en el espacio de características

---

### ✅ Conclusión:

> **"Una neurona no lineal con función de transferencia sigmoide resuelve el mismo tipo de problemas que un combinador lineal"**
> es **FALSO**, porque **la neurona con sigmoide puede resolver más casos** (como clasificación binaria), mientras que el combinador lineal solo se limita a problemas lineales de regresión.


# Ejercicio 14

✅ **Verdadero.**

---

### 🔍 Explicación:

Un **perceptrón** clásico toma la forma:

$$
y =
\begin{cases}
1 & \text{si } \mathbf{w} \cdot \mathbf{x} + b \geq 0 \\
0 & \text{en otro caso}
\end{cases}
$$

Es decir, utiliza una **función de activación escalón** (step function) para producir una salida binaria.

---

Una **neurona no lineal con función de activación sigmoide**:

$$
y = \sigma(\mathbf{w} \cdot \mathbf{x} + b)
\quad \text{donde} \quad \sigma(z) = \frac{1}{1 + e^{-z}}
$$

produce un **valor continuo entre 0 y 1**, pero se puede utilizar **para los mismos fines de clasificación binaria**, interpretando su salida como una **probabilidad**.

---

### ⚖️ Comparación:

| Característica               | Perceptrón              | Neurona con sigmoide       |
| ---------------------------- | ----------------------- | -------------------------- |
| Salida                       | 0 o 1                   | valor continuo en (0, 1)   |
| Activación                   | Escalón (step)          | Sigmoide                   |
| Capacidad para clasificación | Solo problemas lineales | También lineales           |
| Entrenamiento (derivable)    | No                      | Sí ✅ (ideal para backprop) |

---

### ✅ Conclusión:

> **Una neurona no lineal con función de transferencia sigmoide puede resolver el mismo tipo de problema que un perceptrón.**

Es **VERDADERO**: de hecho, es más flexible y entrenable, especialmente con gradientes.


# Ejercicio 15

❌ **Falso.**

---

### 🔍 Explicación:

Una **única neurona con función de activación sigmoide** \*\*solo puede resolver problemas de clasificación binaria cuando las clases son **linealmente separables**.

---

### ⚠️ ¿Por qué?

Porque la salida de la neurona está determinada por una combinación lineal de las entradas:

$$
y = \sigma(\mathbf{w} \cdot \mathbf{x} + b)
$$

Esto define una **frontera de decisión lineal** en el espacio de entrada. Por lo tanto:

* Si las clases se pueden separar con una **recta (en 2D)**, un **plano (en 3D)**, etc., la neurona puede resolver el problema.
* Si las clases **no son linealmente separables** (ej: problema XOR), una sola neurona **no alcanza**.

---

### ✅ ¿Qué se necesita para casos no lineales?

* Una **red neuronal con múltiples capas ocultas** (MLP) puede **aproximar funciones no lineales arbitrarias**.
* Se requiere **composición de neuronas** para lograr **fronteras no lineales**.

---

### 🧠 Ejemplo clásico:

* El problema **XOR** **no puede resolverse** con una sola neurona, aunque tenga activación sigmoide.

---

### ✅ Conclusión:

> **“Puede utilizarse una neurona no lineal con función de activación sigmoide para resolver cualquier problema de clasificación de dos clases.”**
> es **FALSO**, porque sólo puede resolver problemas **linealmente separables**.


# Ejercicio 16

✅ **Verdadero (con matices).**

---

### 🔍 Explicación técnica:

Tanto la función **sigmoide** como la función **tangente hiperbólica (tanh)** son funciones de activación **suaves y no lineales**, y ambas permiten que una neurona:

* Modele **relaciones no lineales**.
* Sea entrenada mediante **descenso del gradiente** (ambas son derivables).
* Se use en tareas como **clasificación binaria**, regresión no lineal, etc.

---

### 📐 Diferencias en práctica:

| Característica  | Sigmoide $\sigma(x)$               | Tanh $\tanh(x)$                  |
| --------------- | ---------------------------------- | -------------------------------- |
| Rango de salida | (0, 1)                             | (−1, 1)                          |
| Simetría        | No (asimétrica)                    | Sí (simétrica respecto a 0)      |
| Uso común       | Salida de modelos de clasificación | Capas ocultas en redes profundas |

---

### ⚠️ ¿Importa la diferencia?

* En teoría, **una sola neurona con sigmoide o tanh puede resolver el mismo tipo de problemas** (por ejemplo, una separación lineal).
* En la práctica, **tanh puede acelerar el entrenamiento** en redes profundas por su simetría.

---

### ✅ Conclusión:

> **“Una neurona no lineal con función de transferencia sigmoide entre 0 y 1 resuelve el mismo tipo de problemas que una neurona no lineal con función de transferencia tanh entre -1 y 1.”**
> es **VERDADERO**, ya que ambas son funciones suaves y no lineales capaces de resolver el mismo tipo de problemas con una sola neurona.

¿Querés ver esto comparado gráficamente o con un ejemplo práctico en código?


# Ejercicio 17

❌ **Falso.**

---

### 🔍 Explicación:

Un **combinador lineal** tiene la forma:

$$
y = \mathbf{w} \cdot \mathbf{x} + b
$$

Para clasificación binaria, suele usarse un **umbral**:

$$
\text{clase} =
\begin{cases}
1 & \text{si } y \geq 0 \\
0 & \text{si } y < 0
\end{cases}
$$

Esto define una **frontera de decisión lineal** (una recta, plano, hiperplano).

---

### ⚠️ Limitación crítica:

Un combinador lineal \*\*solo puede resolver problemas de clasificación binaria si las clases son **linealmente separables**.

---

### 🧠 Ejemplo:

* ✅ Puede resolver: distinguir entre dos clases separadas por una recta (ej. puntos arriba vs. abajo).
* ❌ No puede resolver: el problema **XOR**, donde las clases no se pueden separar con una línea.

---

### ✅ Conclusión:

> **“Puede utilizarse un combinador lineal para resolver cualquier problema de clasificación de dos clases.”**
> es **FALSO**, porque solo funciona cuando las clases son **linealmente separables**. Para casos no lineales, se necesita una red más compleja.


# Ejercicio 18

✅ **Verdadero.**

---

### 🔍 Explicación:

La **función softmax** se usa generalmente para clasificación **multiclase**, y tiene la forma:

$$
\text{softmax}(z_i) = \frac{e^{z_i}}{\sum_{j} e^{z_j}}
$$

Pero cuando se aplica sobre una **sola neurona** (es decir, un único valor $z$), se cumple:

$$
\text{softmax}(z) = \frac{e^z}{e^z} = 1
$$

Entonces **no tiene sentido usar softmax con una sola neurona** de salida de manera directa.

---

### ⚠️ Lo que sí es válido:

En **clasificación binaria**, se suelen usar **dos neuronas en la capa de salida con softmax** (una por clase). Esto sí aplica correctamente:

$$
\text{output} = \text{softmax}([z_1, z_2]) =
\left[ \frac{e^{z_1}}{e^{z_1} + e^{z_2}}, \frac{e^{z_2}}{e^{z_1} + e^{z_2}} \right]
$$

Alternativamente, **se puede usar una sola neurona con activación sigmoide**, que es funcionalmente equivalente a usar softmax sobre 2 clases.

---

### 🧠 Conclusión final:

> **“Puede utilizarse un modelo con una capa de salida softmax formada por una sola neurona para resolver un problema de clasificación binaria.”**

es **✅ VERDADERO** **si interpretamos que hay una segunda clase implícita** (es decir, que se usa softmax para producir una probabilidad sobre dos clases, y solo se retorna una de ellas). Pero en la práctica se prefiere:

* **Softmax con 2 neuronas** para binaria multiclase explícita.
* **Sigmoide con 1 neurona** para binaria estándar.

---

¿Querés que te muestre esto con un ejemplo numérico o gráfico?


# Ejercicio 19

In [ ]:
import pandas as pd      # para trabajar con archivos de datos csv, excel, etc: https://pandas.pydata.org/docs/getting_started/tutorials.html
import chardet           # para detectar la codificación de texto en archivos
import numpy as np

from sklearn import preprocessing, model_selection
from matplotlib import pyplot as plt

from ClassRNMulticlase import RNMulticlase

nombre_archivo = DATOS_DIR + 'Vinos.csv' # archivo de hojas

#-- detectando la codificación de caracteres usada ----
with open(nombre_archivo, 'rb') as f:
    result = chardet.detect(f.read())  # or readline if the file is large

# recupera el archivo en un objeto dataframe de pandas utilizando la codificación detectada
df = pd.read_csv(nombre_archivo, encoding=result['encoding'], sep=';')
df

##### Selección de atributos y normalización

In [ ]:
# %% separa atributos para entrenar de clases
X_raw = np.array(df.iloc[:,1::])  # recupera todas las columnas salvo la primera (es la clase)
T_raw = np.array(df.iloc[:,0])    # recupera solo la primera columna (es la clase)

clases = np.unique(T_raw)  # obtiene las clases sin repeticiones
print('Las clases del dataset son :', clases)

# Normalizacion con media y desviacion
scaler = preprocessing.StandardScaler()
#scaler = preprocessing.MinMaxScaler()
X = scaler.fit_transform(X_raw)

T_raw = T_raw.reshape((-1,1))   # fuerza formato (N,1) sino es (N,)

# los target deben tener formato one hot.
# clase 1 => 100 | clase 2 => 010 | clase 3 => 001
encoder = preprocessing.OneHotEncoder(handle_unknown='ignore')
T = encoder.fit_transform(T_raw.reshape(-1,1)).toarray()


##### Entrenamiento de neurona multiclase con activación "tanh" y error "EC_binaria"

In [ ]:
# Texto para acumular los resultados en formato de tabla
# Se imprime al final de cada experimento con los resultados acumulados
encabezado = "|{:^7}|{:^5}|{:^8}|{:^7}|{:^7}|{:^10}|{:6}|{:6}|{:^8}|{:^8}|".format(
    "div E/P", "ALFA", "FUN", "MAX_ITE", "ERROR", "COSTO",
    "Entren", "Prueba", "100% Ent", "100% Prb"
)
separador = "|{:-^7}|{:-^5}|{:-^8}|{:-^7}|{:-^7}|{:-^10}|{:-^6}|{:-^6}|{:-^8}|{:-^8}|".format(
    "", "", "", "", "", "", "", "", "", "", ""
)

tabla_resultados = encabezado + "\n" + separador + "\n"

In [ ]:
###############################################################################
# neurona multiclase con sigmoid y ECM
ALFA = 0.2       # tasa de aprendizaje
#FUN = 'sigmoid'  # funciones de activacion
FUN = 'tanh'
#FUN = 'softmax'
MAX_ITE = 400    # cant maxima de iteraciones
ERROR = 0.00001    # cota para considerar entrenado el modelo
CORRIDAS = 30    # veces para repetir el experimento
TEST_SIZE = .4   # para division en entrenamiento y prueba
COSTO = 'EC_binaria'    # funciones de costo
#COSTO = 'ECM' | 'EC' | 'EC_binaria'

print('\nNeurona con %s\n' %FUN)

prom_ite = 0
prom_acc_train = 0
prom_acc_test = 0
veces_train = 0
veces_test = 0

for nro_ite in range(CORRIDAS):
    # Establece resultado esperado según la clase. 1=clase esperada, 0=otra clase

    #Importante: la salida de la función de activación debe estar en la misma escala que la clase a predecir
    if FUN == 'tanh':
        T_new = 2*T-1
    else:
        T_new = T

    # genera diferentes divisiones del dataset en cada iteración
    X_train, X_test, Y_train, Y_test = model_selection.train_test_split(X, T_new, test_size=TEST_SIZE, random_state=nro_ite)

    # Entrena modelo
    modelo = RNMulticlase(alpha=ALFA, n_iter=MAX_ITE, FUN=FUN, COSTO=COSTO, cotaE=ERROR)
    modelo.fit(X_train, Y_train)

    (W, b, iteraciones) = (modelo.w_, modelo.b_, len(modelo.errors_))

    prom_ite = prom_ite + iteraciones

    # efectividad entrenamiento
    efectividad = 100*modelo.accuracy(X_train, Y_train)

    prom_acc_train = prom_acc_train + efectividad
    if (100-efectividad < 0.00001):
      veces_train = veces_train + 1

    print("Entrenamiento: %6.2f%% de efectividad con %2d iteraciones" % (efectividad, iteraciones ))
    # efectividad testeo
    efectividad = 100*modelo.accuracy(X_test, Y_test)

    prom_acc_test = prom_acc_test + efectividad
    if (100-efectividad < 0.00001):
      veces_test = veces_test + 1

    print("Prueba.......: %6.2f%% de efectividad con %2d iteraciones" % (efectividad, iteraciones ))

print('-'*80)
print('promedio iteraciones  : %6.2f' % (prom_ite/CORRIDAS))
print('promedio acc. train   : %6.2f' % (prom_acc_train/CORRIDAS))
print('promedio acc. test    : %6.2f' % (prom_acc_test/CORRIDAS))
print('ejec. train con 100%%  : %3d' % (veces_train))
print('ejec.  test con 100%%  : %3d' % (veces_test))


# Añadir fila a la tabla con parámetros y resultados
fila = "|{:^7}|{:^5.1f}|{:^8}|{:^7}|{:^7.0e}|{:^10}|{:6.2f}|{:6.2f}|{:^8}|{:^8}|".format(
    TEST_SIZE,
    ALFA,
    FUN,
    MAX_ITE,
    ERROR,
    COSTO,
    prom_acc_train//CORRIDAS,
    prom_acc_test//CORRIDAS,
    veces_train,
    veces_test
)
tabla_resultados += fila + "\n"

# Imprimir tabla completa
print("\n" + "="*120)
print("RESULTADOS DETALLADOS DE TODAS LAS CORRIDAS:\n")
print(tabla_resultados)

El archivo **`Vinos.csv`** tiene todos los datos en una única columna con separador `;`, por lo que vamos a cargarlo correctamente primero. Luego, evaluamos la afirmación sobre usar una **capa de salida con 3 neuronas tanh + entropía cruzada binaria**.


El dataset contiene:

* **14 columnas**, de las cuales la variable **`Class`** es la clase a predecir.
* **3 clases diferentes**: `[1, 2, 3]` → Es un problema de **clasificación multiclase (3 clases).**

---

### ❓ Afirmación:

> “Es posible clasificar las muestras con una red que tenga una **capa de salida formada por 3 neuronas con activación tanh** y **función de costo entropía cruzada binaria**.”

---

### ❌ **FALSO.**

---

### 🔍 Justificación:

* **`tanh`** produce salidas en el rango **(-1, 1)**, pero:

  * La **entropía cruzada binaria** espera salidas en el rango **(0, 1)** (como las que da una sigmoide o softmax).
  * Además, **entropía cruzada binaria** se usa para **clasificación binaria o multilabel**, no para multiclase con softmax-style outputs.

---

### ✅ Lo correcto sería:

Para este problema de **clasificación multiclase con 3 clases**:

* **Capa de salida**: 3 neuronas con **activación softmax**.
* **Función de pérdida**: **categorical crossentropy** (o sparse categorical crossentropy si las clases son enteros).

---

### 🧠 Conclusión:

> “Es posible clasificar las muestras con una red con 3 neuronas `tanh` y pérdida `binary crossentropy`.”

Es **FALSO**, porque:

* El **rango de activación y la función de pérdida no son compatibles**.
* El problema es multiclase, no multilabel ni binario.

¿Querés que te dé el código correcto en Keras para este caso?


# Ejercicio 20

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from numpy import tanh

# Cargar el archivo
df = pd.read_csv("FrutasTrain.csv")  # Asegurate de ajustar el path si es necesario

# Extraer características
X = df[['Diametro', 'Color']].values

# Normalizar con StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Pesos dados de la neurona entrenada
W = np.array([2.40, -2.39])
b = -0.065

# Calcular z = W·X + b
z = X_scaled @ W + b

# Aplicar activación tanh
salidas = tanh(z)

# Clasificación:
# - Si salida ≥ 0.8 → clase 1
# - Si salida ≤ -0.8 → clase -1
# - Si salida ∈ (-0.8, 0.8) → indefinido
indefinidos = np.sum((salidas > -0.8) & (salidas < 0.8))

print("Cantidad de ejemplos indefinidos:", indefinidos)


✅ La cantidad de ejemplos de entrenamiento clasificados como **“indefinido”** es:

**🔢 5 ejemplos**

---

### 📌 Según el criterio:

* Salida ≥ 0.8 → clase **1**
* Salida ≤ -0.8 → clase **−1**
* Salida ∈ (−0.8, 0.8) → **indefinido**

---

### ✅ Opción correcta:

**f. 5**


# Ejercicio 21

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from numpy import tanh

# Cargar los datos de entrenamiento
df = pd.read_csv("FrutasTrain.csv")  # Asegurarse que el archivo esté en el path correcto

# Extraer características numéricas
X = df[['Diametro', 'Color']].values

# Normalizar los datos de entrenamiento
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Definir los pesos de la neurona entrenada
W = np.array([2.40, -2.39])
b = -0.065

# Nueva fruta: diámetro = 14, color = 140
nueva_fruta = np.array([[14, 140]])

# Aplicar la misma normalización
nueva_fruta_scaled = scaler.transform(nueva_fruta)

# Calcular salida de la neurona
z_nueva = nueva_fruta_scaled @ W + b
salida_nueva = tanh(z_nueva[0])

# Clasificar según el criterio
if salida_nueva >= 0.8:
    resultado = "Naranja"
elif salida_nueva <= -0.8:
    resultado = "Melón"
else:
    resultado = "Indefinido"

# Mostrar resultado
print(f"Salida de la neurona: {salida_nueva:.4f}")
print(f"Clasificación: {resultado}")


✅ La salida de la neurona para una fruta con:

* **Diámetro = 14 cm**
* **Color = 140**

es aproximadamente **−0.97**

---

### 📌 Clasificación:

* Como la salida es **menor o igual que −0.8**, se clasifica como:

**🥭 Melón**

---

### ✅ Opción correcta:

**b. Melón**


# Ejercicio 22

In [ ]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from numpy import tanh

# Cargar datos de entrenamiento
df = pd.read_csv("FrutasTrain.csv")  # Asegurate de que el archivo esté en el mismo directorio

# Extraer características
X = df[['Diametro', 'Color']].values

# Normalizar los datos con StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Definir pesos de la neurona entrenada
W = np.array([2.40, -2.39])
b = -0.065

# Nueva fruta a clasificar: diámetro = 16, color = 79
nueva_fruta = np.array([[16, 79]])

# Normalizar la nueva muestra con el mismo scaler
nueva_fruta_scaled = scaler.transform(nueva_fruta)

# Calcular la salida de la neurona
z = nueva_fruta_scaled @ W + b
salida = tanh(z[0])

# Clasificar según criterio definido
if salida >= 0.8:
    resultado = "Naranja"
elif salida <= -0.8:
    resultado = "Melón"
else:
    resultado = "Indefinido"

# Mostrar resultados
print(f"Salida de la neurona: {salida:.4f}")
print(f"Clasificación: {resultado}")

# Salida de la neurona: 0.5918
# Clasificación: Indefinido


✅ La salida de la neurona para una fruta con:

* **Diámetro = 16 cm**
* **Color = 79**

es aproximadamente **0.59**

---

### 📌 Clasificación:

Como la salida está en el rango **(−0.8, 0.8)**, la clasificación es:

**🚫 Indefinido**

---

### ✅ Opción correcta:

**b. Indefinido**
